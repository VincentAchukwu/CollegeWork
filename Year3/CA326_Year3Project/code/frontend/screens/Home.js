import React, {Component, useState} from "react";
import { StyleSheet, FlatList, Text, View, StatusBar, TextInput, TouchableOpacity } from "react-native";
import LinearGradient from "react-native-linear-gradient";

// specifying URL of Flask server
const flaskAPI = "https://currencyapp200.pythonanywhere.com/api/conversions?"

const Home = () => {

    const [fromCurrency, setFromCurrency] = useState("");
    const [fromCurrencyError, setFromCurrencyError] = useState(null);
    const [toCurrency, setToCurrency] = useState("");
    const [toCurrencyError, setToCurrencyError] = useState(null);
    const [amount, setAmount] = useState(0);
    const [amountError, setAmountError] = useState(null);
    const [result, setResult] = useState(null);
    // For testing
    const [status, setStatus] = useState("Press");

    // checks if the "convert from" field is not empty
    const fromCurrencyValidator = () => {
        if(fromCurrency == ""){
            setFromCurrencyError("field cannot be empty!")
        }
        else{
            setFromCurrencyError("")
        }
    }

    // checks if the "convert to" field is not empty
    const toCurrencyValidator = () => {
        if(toCurrency == ""){
            setToCurrencyError("field cannot be empty!")
        }
        else{
            setToCurrencyError("")
        }
    }

    // checks if the "amount" field is not empty
    const amountValidator = () => {
        if(amount == ""){
            setAmountError("field cannot be empty!")
        }
        else{
            setAmountError("")
        }
    }

    return (
        <View style={styles.container}>
            <StatusBar backgroundColor="#1E90FF" barStyle="light-content"/>

            <View style= {[styles.row, {marginTop: "10%"}]}>
                <View style= {[styles.action, {margin: "3%"}]}>
                    <TextInput
                        testID="convertFrom"
                        style= {{flex:1}}
                        maxLength={3}
                        onBlur={()=> fromCurrencyValidator()}
                        placeholder="e.g EUR"
                        onChangeText={(text) => setFromCurrency(text)}
                    />
                </View>
                <Text style= {{fontSize: 18, marginTop: "5%", fontWeight: "bold"}}> To: </Text>
                <View style= {[styles.action, {margin: "3%"}]}>
                    <TextInput
                        testID="convertTo"
                        style= {{flex:1}}
                        maxLength={3}
                        onBlur={()=> toCurrencyValidator()}
                        placeholder="e.g GBP"
                        onChangeText={(text) => setToCurrency(text)}
                    />
                </View>
            </View>

            <View style= {styles.center}>
                <Text style={[styles.emptyError, {marginTop: "2%"}]}>{fromCurrencyError}</Text>
                <Text style={styles.emptyError}>{toCurrencyError}</Text>

                <Text style={ styles.footerText }> Amount: </Text>
                <View style= {styles.action}>
                    <TextInput
                        testID="amount"
                        style= {{flex:1}}
                        onBlur={() => amountValidator()}
                        keyboardType="numeric"
                        placeholder="e.g 25"
                        onChangeText={(text) => setAmount(text)}
                    />
                </View>
                
                <Text style={styles.emptyError}>{amountError}</Text>

                <TouchableOpacity
                    style={styles.convert}
                    disabled={!fromCurrency || !toCurrency || !amount}
                    onPress={async () => {
                        const convertEndoint = "fromCurr="+fromCurrency+"&toCurr="+toCurrency+"&amount="+amount;
                        const response = await fetch(flaskAPI+convertEndoint, {
                            method: "GET",
                            mode: "CORS",
                        })
                        .then(res => res.json())
                        .then(data => {
                            setResult(data.requestedConversion)
                        });
                    }}
                >
                    <LinearGradient
                        colors={ ["#09C6F9", "#045DE9"] }
                        style={ styles.button }
                        >
                        <Text style={styles.convertText}>Convert</Text>
                    </LinearGradient>
                </TouchableOpacity>

                <View style = {styles.lineStyle} />
                
                {/* basically the flatlist shows result of each currency rate
                unless the input was invalid, it'll display error message */}

                {/* Button Testing */}
                {/*<TouchableOpacity
                    testID="myButton"
                    onPress= {() => setStatus("button pressed")}
                >
                    <Text testID="myText">{status}</Text>
                </TouchableOpacity>*/}

                <View style={[styles.action, {marginTop: "8%"}] }>
                    <FlatList
                        data={result}
                        keyExtractor={item => item.key}
                        renderItem={({ item }) => (
                            <View>
                                {item.key != "invalid" ? <Text>{toCurrency.toUpperCase()} = {item.key}</Text> : <Text style={{color: "red"}}>Please enter valid currency codes and amount</Text> }
                            </View>
                        )}
                    />
                </View>
            </View>
        </View>
    )

}


const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: "#ffffff"
    },
    footerText: {
        fontSize: 18,
        paddingLeft: "3%"
    },
    row: {
        flexDirection: "row",
        justifyContent: "center",
    },
    input: {
        borderWidth: 1,
        padding: 8,
        margin: 10,
        width: 200,
    },
    emptyError: {
        color: "red",
    },
    action: {
        flexDirection: 'row',
        alignItems: 'center',
        borderWidth: .8,
        height: 50,
        borderRadius: 30,
        paddingLeft: "4%",
        paddingRight: "4%",
        margin: "3%",
        width:"35%"
    },
    button: {
        width: '100%',
        height: 50,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 30
    },
    convert: {
        width: '40%',
        height: 50,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 30,
        marginTop: "4%"
    },
    convertText: {
        fontSize: 18,
        textAlign: "center",
        color: "white",
    },
    center: {
        justifyContent: "center",
        alignItems: "center",
        marginTop: "-5%"
    },
    lineStyle:{
        borderBottomWidth: 10,
        borderRadius: 30,
        width: "100%",
        paddingTop: "5%",
        paddingBottom: "5%",
        borderBottomColor: "#1E90FF",
    },
})

export default Home;