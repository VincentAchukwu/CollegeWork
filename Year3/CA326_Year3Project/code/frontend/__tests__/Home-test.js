import "react-native";
import React from "react";
import Home from "../screens/Home";
import renderer from "react-test-renderer";
import {create, act} from "react-test-renderer"

const tree = create(<Home/>);

// ********* Screen Rendering Test *********

test("Home Snapshot", () =>{
    const snap = renderer.create(
        <Home/>
    ).toJSON;

expect(snap).toMatchSnapshot();
})

// *****************************************

// ********* Button Testing *********

test("button press", () => {
    // press button
    const button = tree.root.findByProps({ testID: "myButton"}).props;
    act(() => button.onPress());

    // express text to equal "button pressed"
    const text = tree.root.findByProps({ testID: "myText"}).props;
    expect(text.children).toEqual("button pressed")
})

// ***********************************

// ********* Function to Find Elements in a Tree (Tree being <View/> component and everything wrapped in it) *********

const findElement = function(tree2, element) {
    let result = undefined
    for(node in tree2.children){
        if(tree2.children[node].props.testID = element) {
            result = true;
        }
    }
    return result
};

// *******************************************************************************************************************

// ********* TextInput Tests *********

it("Has Element convertFrom", () => {
    const tree2 = renderer.create(<Home/>).toJSON();
    expect(findElement(tree2, "convertFrom")).toBeDefined();
});

it("Has Element convertTo", () => {
    const tree2 = renderer.create(<Home/>).toJSON();
    expect(findElement(tree2, "convertTo")).toBeDefined();
});

it("Has Element amount", () => {
    const tree2 = renderer.create(<Home/>).toJSON();
    expect(findElement(tree2, "amount")).toBeDefined();
});

// ***********************************