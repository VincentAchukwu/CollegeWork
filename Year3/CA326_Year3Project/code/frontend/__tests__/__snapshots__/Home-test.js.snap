// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Snapshot 1`] = `[Function]`;
